plugins/ipam/ipam.go

```diff
@@ -1036,29 +1147,185 @@ func (i *IPAM) SidForNodeToNodeHostLocalsid(nodeIP net.IP) net.IP {
+}
+
+// BsidForSFCPolicy creates a valid SRv6 SID for policy used for SFC
+func (i *IPAM) BsidForSFCPolicy(sfcName string) net.IP {
+	// prepare computation values
+	prefix := i.ContivConf.GetIPAMConfig().SRv6Settings.SFCPolicyBSIDSubnetCIDR
+	prefixMaskSize, _ := prefix.Mask.Size()
+	sfcID := i.computeSFCID(sfcName)
+
+	// compute BSID as combination of configurable prefix and SFC ID
+	return i.combineMultipleIPAddresses(
+		newIPWithPositionableMaskFromIPNet(prefix),
+		newIPWithPositionableMask(sfcID, prefixMaskSize, 128-prefixMaskSize))
+}
+
+// SidForSFCExternalIfLocalsid creates a valid SRv6 SID for external interface
+func (i *IPAM) SidForSFCExternalIfLocalsid(externalIfName string, externalIfIP net.IP) net.IP {
+	var ip net.IP
+	if externalIfIP != nil {
+		ip = i.SidForSFCEndLocalsid(externalIfIP)
+	} else {
+		ip = i.computeExtIfID(externalIfName)
+	}
+
+	prefix := i.ContivConf.GetIPAMConfig().SRv6Settings.SFCEndLocalSIDSubnetCIDR
+	prefixMaskSize, _ := prefix.Mask.Size()
+	return i.combineMultipleIPAddresses(
+		newIPWithPositionableMaskFromIPNet(prefix),
+		newIPWithPositionableMask(ip, prefixMaskSize, 128-prefixMaskSize))
+}
+
+// SidForSFCServiceFunctionLocalsid creates a valid SRv6 SID for locasid leading to pod of service function given by
+// <serviceFunctionPodIP> IP address.
+func (i *IPAM) SidForSFCServiceFunctionLocalsid(sfcName string, serviceFunctionPodIP net.IP) net.IP {
+	// prepare computation values
+	prefix := i.ContivConf.GetIPAMConfig().SRv6Settings.SFCServiceFunctionSIDSubnetCIDR
+	prefixMaskSize, _ := prefix.Mask.Size()
+	sfcID := i.computeSFCID(sfcName)
+	sfcIDMaskLength := int(i.ContivConf.GetIPAMConfig().SRv6Settings.SFCIDLengthUsedInSidForServiceFunction)
+
+	// compute SID as combination of configurable prefix, SFC ID and IP address of service function pod
+	return i.combineMultipleIPAddresses(
+		newIPWithPositionableMaskFromIPNet(prefix),
+		newIPWithPositionableMask(sfcID, prefixMaskSize, sfcIDMaskLength),
+		newIPWithPositionableMask(serviceFunctionPodIP, prefixMaskSize+sfcIDMaskLength,
+			128-prefixMaskSize-sfcIDMaskLength))
+}
+
+// SidForSFCEndLocalsid creates a valid SRv6 SID for locasid of segment that is the last link of SFC chain
+func (i *IPAM) SidForSFCEndLocalsid(serviceFunctionPodIP net.IP) net.IP {
+	// prepare computation values
+	prefix := i.ContivConf.GetIPAMConfig().SRv6Settings.SFCEndLocalSIDSubnetCIDR
+	prefixMaskSize, _ := prefix.Mask.Size()
+
+	// compute SID as combination of configurable prefix and IP address of service function pod
+	return i.combineMultipleIPAddresses(
+		newIPWithPositionableMaskFromIPNet(prefix),
+		newIPWithPositionableMask(serviceFunctionPodIP, prefixMaskSize, 128-prefixMaskSize))
+}
+
+// UpdateExternalInterfaceIPInfo is notifying IPAM about external interfacew IP allocation
+func (i *IPAM) UpdateExternalInterfaceIPInfo(extif, vppInterface string, nodeID uint32, ipNet *net.IPNet,
```

Comment:
Couldn't IPAM read (and resync) IP addresses of external interfaces like IPNet plugin does from `kubeStateData[extifmodel.Keyword]`?
This would help to get rid of the Update... method from the IPAM plugin API.
Ideally, the IPAM plugin should only have Allocate/Get methods in its API.

Reply:
Done here [https://github.com/contiv/vpp/pull/1663/commits/f1dff08f7085a916ddaf28b47fab261b7a039432](url)